erpnext/accounts/doctype/budget/budget.py

```diff
@@ -27,12 +27,19 @@ def validate(self):
 	def validate_duplicate(self):
 		budget_against_field = frappe.scrub(self.budget_against)
 		budget_against = self.get(budget_against_field)
-		existing_budget = frappe.db.get_value("Budget", {budget_against_field: budget_against,
+		# get existing 'budget' doctype
+		for existing_budget in frappe.db.get_values("Budget", {budget_against_field: budget_against,
 			"fiscal_year": self.fiscal_year, "company": self.company,
-			"name": ["!=", self.name], "docstatus": ["!=", 2]})
-		if existing_budget: 
-			frappe.throw(_("Another Budget record '{0}' already exists against {1} '{2}' for fiscal year {3}")
-				.format(existing_budget, self.budget_against, budget_against, self.fiscal_year), DuplicateBudgetError)
+			"name": ["!=", self.name], "docstatus": ["!=", 2]},["name"], as_dict=1):
+			# get 'budget account' child table for existing budget doctype
+			for budget_account_details in frappe.db.get_values("Budget Account",
+			{"parent": existing_budget.name,"parenttype":"Budget"}, ["account"], as_dict=1):	
+				if budget_account_details.account:
+					# get 'account' from current doctype
+					for present_account in self.get('accounts'):
+						if present_account.account == budget_account_details.account:
+							frappe.throw(_("Another Budget record '{0}' already exists against {1} '{2}' for account '{3}' & fiscal year {4}")
+								.format(existing_budget.name, self.budget_against, budget_against, present_account.account,self.fiscal_year), DuplicateBudgetError)
```

Comment:
Instead of multiple fir loops, you can directly check using join query between doctype 'Budget' and 'Budget Account' and if records found then throw an exception
like frappe.db.sql(""" check record found against the Cost Center/Project + Account+ Fiscal Year """)